sklearn/mixture/gaussian_mixture.py

```diff
@@ -237,14 +263,21 @@ def _estimate_gaussian_covariance_spherical(resp, X, nk, means, reg_covar):
 
     Returns
     -------
-    covariances : array, shape (n_components,)
+    precisions : array, shape (n_components,)
     """
-    covariances = _estimate_gaussian_covariance_diag(resp, X, nk, means,
-                                                     reg_covar)
-    return covariances.mean(axis=1)
+    avg_X2 = np.dot(resp.T, X * X) / nk[:, np.newaxis]
+    avg_means2 = means ** 2
+    avg_X_means = means * np.dot(resp.T, X) / nk[:, np.newaxis]
+    covariances = (avg_X2 - 2 * avg_X_means + avg_means2 + reg_covar).mean(1)
+    if np.any(np.less_equal(covariances, 0.0)):
+        raise ValueError("The algorithm has diverged because of too "
+                         "few samples per components. "
+                         "Try to decrease the number of components, or "
+                         "increase reg_covar.")
```

Comment:
Please put this message in a module level constant instead of repeating it several times for each covariance type.